File: C_repo/test_9.c
#include <stdio.h>

// Tests 'void*' pointers, which are C's generic pointer type.
// This is very hard to translate to strongly-typed Java.

struct TypeA {
    int value;
};

struct TypeB {
    double value;
};

void printGeneric(void* data, int type) {
    if (type == 1) {
        struct TypeA* a = (struct TypeA*)data;
        printf("TypeA: %d\n", a->value);
    } else if (type == 2) {
        struct TypeB* b = (struct TypeB*)data;
        printf("TypeB: %f\n", b->value);
    }
}

int main() {
    struct TypeA a = {123};
    struct TypeB b = {456.789};
    
    printGeneric(&a, 1);
    printGeneric(&b, 2);
    return 0;
}

File: Runs/Java_repo_gemma3:12b/Test_9.java
public class Test_9 {

    public static class TypeA {
        public int value;
    
        public TypeA() {
        }
    
        public TypeA(TypeA other) {
            this.value = other.value;
        }
    
        public TypeA(int value) {
            this.value = value;
        }
    }

    public static class TypeB {
        public double value;
    
        public TypeB() {
        }
    
        public TypeB(TypeB other) {
            this.value = other.value;
        }
    
        public TypeB(double value) {
            this.value = value;
        }
    }

    public static void printGeneric(Object data, int type) {
        if (type == 1) {
            TypeA a = (TypeA) data;
            System.out.printf("TypeA: %d\n", a.value);
        } else if (type == 2) {
            TypeB b = (TypeB) data;
            System.out.printf("TypeB: %f\n", b.value);
        }
    }

    public static void printGeneric(Object a, int type) {
        if (type == 1) {
            TypeA ta = (TypeA) a;
            System.out.println("TypeA value: " + ta.value);
        } else if (type == 2) {
            TypeB tb = (TypeB) a;
            System.out.println("TypeB value: " + tb.value);
        }
    }

}